Improvements to the TimeSeries guide (#6413)
sunpy/sunpy: docs/guide/data_types/timeseries.rst

.. _timeseries_guide:

**********************
Time series data guide
**********************

Time series data are a fundamental part of many data analysis projects in heliophysics as well as other areas.
**sunpy** provides a TimeSeries object to handle this type of data.
Much like the `~sunpy.map.Map` object, `~sunpy.timeseries.TimeSeries` can load generic filetypes, and recognizes data from nine specific sources to provide instrument specific data loading and plotting capabilities.

> For more information about TimeSeries, and what file types and data sources are supported, check out :doc:`/code_ref/timeseries`.

Reviewer: ```suggestion
For more information about `~sunpy.timeseries.TimeSeries`, and what file types and data sources are supported, check out :doc:`/code_ref/timeseries`.
```

Reviewer: Now, that I'm looking down the list, I see that your subsequent mentions of `TimeSeries` are not linked either. If that's intentional, I'm fine with that and feel free to ignore this suggestion. Perhaps it is too distracting to the eye to have every mention of of the object linked.

Author: I didn't really think about this too much, and think I might leave it for this PR, but thanks for catching - I might go through and intentionally link/not-link in another PR.

    >>> import sunpy.data.sample  # doctest: +REMOTE_DATA
    >>> my_timeseries = ts.TimeSeries(sunpy.data.sample.GOES_XRS_TIMESERIES)  # doctest: +REMOTE_DATA

This is calling the `~sunpy.timeseries.TimeSeries` factory to create a time series from a GOES XRS FITS file.

The variable ``my_timeseries`` is a `~sunpy.timeseries.GenericTimeSeries` object.
To create one from a local GOES/XRS FITS file try the following:

.. code-block:: python

    >>> my_timeseries = ts.TimeSeries('/mydirectory/myts.fits', source='XRS')   # doctest: +SKIP

**sunpy** will attempt to detect automatically the instrument source for most FITS files.

> However timeseries data are stored in a variety of file types (FITS, txt, csv, CDF), and so it is not always possible to detect the source.

Reviewer: The usage of "timeseries" vs "time series" is inconsistent. A quick google shows that both are acceptable. I do not have an opinion one way or another, but we should be consistent.

Reviewer: It looks like we mostly use "time series" so maybe it is just easier to stick with that.

Author: 👍  thanks for catching this! I went with "time series" and fixed them all.

**sunpy** ships with a number of known instrumental sources, and can also load CDF files that conform to the `Space Physics Guidelines for CDF <https://spdf.gsfc.nasa.gov/sp_use_of_cdf.html>`__.
If you would like **sunpy** to include another instrumental source see the `Newcomers' Guide <https://docs.sunpy.org/en/latest/dev_guide/contents/newcomers.html>`__.

The `~sunpy.timeseries.TimeSeries` factory has the ability to create a list of TimeSeries objects using a list of filepaths, a folder or a glob, for example:

.. code-block:: python

    >>> my_ts_list = ts.TimeSeries('filepath1', 'filepath2', source='XRS')   # doctest: +SKIP
    >>> my_ts_list = ts.TimeSeries('/goesdirectory/', source='XRS')   # doctest: +SKIP
    >>> my_ts_list = ts.TimeSeries(glob, source='XRS')   # doctest: +SKIP

When manually specifing the source this functionality will only work with files from the same single source, generating a source specific child of the `~sunpy.timeseries.GenericTimeSeries` class such as the `~sunpy.timeseries.sources.goes.XRSTimeSeries` above.

> Instead of creating a list of one TimeSeries object per file, you can create a single time series from multiple files using the keyword argument ``concatenate=True``:

Reviewer: See above re: inconsistent usage of "time series" vs "timeseries"

.. code-block:: python

    >>> from sunpy.time import TimeRange
    >>> tr = TimeRange('2012-06-01 05:00', '2012-06-01 06:30')
    >>> my_timeseries_trunc = my_timeseries.truncate(tr) # doctest: +REMOTE_DATA

This takes a number of different arguments, such as the start and end dates (as datetime or string objects) or a `~sunpy.time.TimeRange` as used above.
Note that the truncated TimeSeries will have a truncated `~sunpy.timeseries.TimeSeriesMetaData` object, which may include dropping metadata entries for data totally cut out from the TimeSeries.
If you want to truncate using slice-like values you can, for example taking every 2nd value from 0 to 10000 can be done using:

.. code-block:: python

    >>> my_timeseries_trunc = my_timeseries.truncate(0, 100000, 2) # doctest: +REMOTE_DATA

> 4.3 More complicated timeseries operations

Reviewer: ```suggestion
4.3 More complicated time series operations
```

------------------------------------------
If you want to do any more complicated analysis on a TimeSeries, we recommend converting it to a `pandas.DataFrame` object first.
Although this conversion will use the unit information and metadata, pandas has a wide array of methods that can be used e.g. for resampling data.
As an example to downsample you can do:

.. code-block:: python

    >>> downsampled_dataframe = my_timeseries_trunc.to_dataframe().resample('10T').mean() # doctest: +REMOTE_DATA

Here ``10T`` means sample every 10 minutes and 'mean' is the method used to combine the data in each 10 minute bin.

> See the `pandas` documentation for more details on other functionality they offer for timeseries analysis.

Reviewer: ```suggestion
See the `pandas` documentation for more details on other functionality they offer for time series analysis.
```
